allennlp/models/bidirectional_lm.py

```diff
@@ -1,57 +1,20 @@
-from typing import Dict, List, Tuple, Union
+from typing import Union
 
-import torch
-import numpy as np
-
-from allennlp.common.checks import ConfigurationError
 from allennlp.data.vocabulary import Vocabulary
+from allennlp.models.language_model import LanguageModel
 from allennlp.models.model import Model
 from allennlp.modules.text_field_embedders import TextFieldEmbedder
-from allennlp.modules.sampled_softmax_loss import SampledSoftmaxLoss
 from allennlp.modules.seq2seq_encoders import Seq2SeqEncoder
-from allennlp.nn.util import get_text_field_mask
 from allennlp.nn import InitializerApplicator
 
 
-class _SoftmaxLoss(torch.nn.Module):
-    """
-    Given some embeddings and some targets, applies a linear layer
-    to create logits over possible words and then returns the
-    negative log likelihood.
-    """
-    def __init__(self,
-                 num_words: int,
-                 embedding_dim: int) -> None:
-        super().__init__()
-
-        # TODO(joelgrus): implement tie_embeddings (maybe)
-        self.tie_embeddings = False
-
-        self.softmax_w = torch.nn.Parameter(
-                torch.randn(embedding_dim, num_words) / np.sqrt(embedding_dim)
-        )
-        self.softmax_b = torch.nn.Parameter(torch.zeros(num_words))
-
-    def forward(self, embeddings: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
-        # pylint: disable=arguments-differ
-        # embeddings is size (n, embedding_dim)
-        # targets is (batch_size, ) with the correct class id
-        # Does not do any count normalization / divide by batch size
-        probs = torch.nn.functional.log_softmax(
-                torch.matmul(embeddings, self.softmax_w) + self.softmax_b,
-                dim=-1
-        )
-
-        return torch.nn.functional.nll_loss(probs, targets.long(), reduction="sum")
-
-
 @Model.register('bidirectional-language-model',
                 deprecation_message=('The "bidirectional-language-model" name was '
                                      'deprecated in version 0.8 and will be removed'
                                      'in version 0.10 . '
```

Comment:
ok, so you don't want to deprecate `bidirectional-language-model`?

Reply:
I don't hold an opinion on whether to apply the deprecation warning or not*. Encouraging standardizing on underscores is reasonable. However, I don't want to actually remove the name as it breaks the API. I think API breakage should be used only when strictly necessary.

* Depends on what we want to communicate as a team with deprecation warnings. I think it can be useful to communicate that "you shouldn't use this there is something better" without also saying "we're definitely going to remove this soon". This varies across projects.

Reply:
thanks for clarifying!